nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,157 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(
```

Comment:
@seanmor5 I've managed to refactor this function for operating directly on binaries and to avoid memory pre-allocation.
All tests continue to pass, so that's something haha

There may be one last trick we can apply on the construction of Q due to the nature of the reflector matrices, but I don't think I can optimize this implementation further.

The main idea behind the latest refactor (aside from operating upon binaries) is that the reflector can be constructed in a single K*K loop, where we can both calculate the reflection and the "identity matrix wrapper" easily